### ISSUE TITLE
Training instability

### ISSUE THREAD

**Issue by cgsas** [mooninrain/FAFE#1]
compute_double_geodesic_error() so3_loss part expereinced training insatbility, despite efforts to clamp the angle (not close +/- 180 deg).

SO(3) loss is notoriously unstable to train -- wonder how you achieved training stability.  

Thanks.

**Comment by mooninrain** [comment]
Hi, thank you for the feedback! I have not experienced instability when fine-tuning AF2 for antibody-antigen docking, but as I stated in the paper, I have not tested my hyperparameter setting in other scenarios, for example, pretraining. Could you provide more information?

Also, could you try to do the following thing: use original FAPE loss(i.e., drop the SO(3) term or simply detach its gradient), but record the SO(3) loss? The SO(3) loss should still decrease even if you do not directly optimize it, because FAPE itself provides a chordal estimation of it :)

**UPDATE: I just double checked the code and here is something that came to my mind.**

> The original FAPE loss is applied with clamping. Specifically, for some samples the loss is only applied to residue pairs
within certain distance. In the early version of F2E, we apply the same restriction to the R(3) part of the loss. However,
later experiments show that removing the gradients of the SO(3) at the same time can stabilize and accelerate the training
procedure.

As I wrote in the supplementary, Later I found that adding extra clamping for SO(3) can stablize training. All the experiments in the paper, however, is done with the original version because I didn't have the time to re-do them when submitting the paper. According to my own ablation this one should be much more stable. Please replace compute_double_geodesic_error() with the following:

```
@tensor_utils.compute_in_fp32
def compute_double_geodesic_error(
    R_pred: Tensor,
    t_pred: Tensor,
    R_gt: Tensor,
    t_gt: Tensor,
    frame_mask: Tensor,
    enable_rotate_loss: bool = True,
    enable_rotate_clamp: bool = True,
    rotate_scale: float = 1.0,
    axis_scale: float = 20.0,
    dist_clamp: float | None = None,
    eps: float = 1e-7,
    group_aware: bool = False,
) -> tuple[base.ProteinFoldingLoss, base.ProteinFoldingError]:
    """Compute frame-wise error with double geodesic metric.

    d_se3(T_pred, T_gt) = √(d_so3(R_pred, R_gt)^2 + (d_r3(t_pred, t_gt) / axis_scale)^2)

    d_so3(R_pred, R_gt) range [0, pi]

    d_r3(t_pred, t_gt) / axis_scale) range [0, 1.5] when clamping

    Reference: https://github.com/airalcorn2/pytorch-geodesic-loss/blob/master/geodesic_loss.py

    Args:
        R_pred: Predicted rotation matrices of T, (..., N, 3, 3).
        t_pred: Predicted translations of T, (..., N, 3).
        R_gt: Ground truth rotation matrices of T, (..., N, 3, 3).
        t_gt: Ground truth translations of T, (..., N, 3).
        frame_mask: Existing masks of ground truth T, (..., N).
        enable_rotate_loss: Bool to enable/disable SO3 loss. When False, fall back to backbone FAPE.
        enable_rotate_clamp: Bool to enable/disable clamping of SO3 loss. Clamp mask is applied according to R3 loss.
        rotate_scale: Scale by which the SO3 part of loss is divided.
        axis_scale: Scale by which the R^3 part of loss is divided.
        dist_clamp: Cutoff above which distance errors are disregarded.
        eps: Small number for numeric safety for arccos.
            Refer to https://github.com/pytorch/pytorch/issues/8069
        group_aware: Bool to use masked quadratic mean(==True) or masked mean(==False).

    Returns:
        Dict of (B) FAFE losses. Contains "fafe", "fafe_so3", "fafe_r3".
        Dict of (B, N) FAFE errors. Contains "fafe", "fafe_so3", "fafe_r3".

    Note:
        so3 loss/error presented in scaled form [0, pi/rotate_scale].
        r3 loss/error presented in scaled form [0, dist_clamp/axis_scale].
    """
    if dist_clamp is None:
        dist_clamp = 1e9
    _mean_func = common.mask.masked_mean if not group_aware else common.mask.masked_quadratic_mean

    # R3 loss
    r3_dist = common.geometric.euclidean_distance(t_pred, t_gt)  # (..., N)
    clamp_mask = r3_dist >= dist_clamp  # used for SO(3) loss clamping
    r3_dist = r3_dist.clamp(max=dist_clamp) / axis_scale  # (..., N)
    r3_loss, r3_error = _mean_func(r3_dist, frame_mask, dim=(-1), return_masked=True)

    # SO3 loss
    # Getting the trace of the matrix product is equivalent to summing the element-wise product of the two matrices
    # with one transposed
    # R_diff = einops.rearrange(R_pred, "... i j -> ... j i") @ R_gt
    # R_diff_trace = R_diff.diagonal(dim1=-2, dim2=-1).sum(-1)
    R_diff_trace = (R_pred * R_gt).sum([-1, -2])
    so3_dist = torch.acos(torch.clamp((R_diff_trace - 1) / 2, -1 + eps, 1 - eps)) / rotate_scale  # (..., N)
    if enable_rotate_clamp:
        so3_dist = torch.where(clamp_mask, so3_dist.detach(), so3_dist)  # stop gradients where r3 is clamped

    so3_loss, so3_error = _mean_func(so3_dist, frame_mask, dim=(-1), return_masked=True)

    # double geodesic loss
    if enable_rotate_loss:
        se3_dist = torch.sqrt(so3_dist**2 + r3_dist**2)  # (..., N)
    else:
        se3_dist = r3_dist  # fall back to FAPE
    se3_loss, se3_error = _mean_func(se3_dist, frame_mask, dim=(-1), return_masked=True)

    losses: base.ProteinFoldingLoss = {
        "fafe": se3_loss,  # Note se3_loss = √((so3_loss/rotate_scale)^2 + (r3_loss/axis_scale)^2)
        "fafe_so3": so3_loss,
        "fafe_r3": r3_loss,
    }

    errors: base.ProteinFoldingError = {
        "fafe": se3_error.detach(),
        "fafe_so3": so3_error.detach(),
        "fafe_r3": r3_error.detach(),
    }
    return losses, errors
```
